pkg/placement/placement-decision.go

```diff
@@ -0,0 +1,173 @@
+	generatedPlacementDecisionSpec, err := c.placementResolver.GeneratePlacementDecision(placementIdentifier)
+	if err != nil {
+		return fmt.Errorf("failed to get placement decision spec: %w", err)
+	}
+
+	// calculate if the resolved decision is different from the current one
+	if !c.placementResolver.ComparePlacementDecision(placementIdentifier, &placementDecision.Spec) {
+		// update the placement decision object in the cluster by updating spec
+		if err = c.updateOrCreatePlacementDecision(placementDecision, generatedPlacementDecisionSpec); err != nil {
+			return fmt.Errorf("failed to update or create placement decision: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// updateOrCreatePlacementDecision updates or creates a placement-decision object in the cluster.
+// If the object already exists, it is updated. Otherwise, it is created.
+func (c *Controller) updateOrCreatePlacementDecision(pd *v1alpha1.PlacementDecision,
+	placementDecisionSpec *v1alpha1.PlacementDecisionSpec) error {
+	unstructuredPlacementDecision, err := placementDecisionSpecToUnstructuredObject(pd, placementDecisionSpec)
+	if err != nil {
+		return fmt.Errorf("failed to update or create placement decision: %w", err)
+	}
+
+	// set owner reference
+	if err = setOwnerReference(c.dynamicClient, unstructuredPlacementDecision); err != nil {
+		return fmt.Errorf("failed to set owner-reference for placement-decision %v: %w", pd.GetName(), err)
+	}
+
+	_, err = c.dynamicClient.Resource(schema.GroupVersionResource{
+		Group:    v1alpha1.SchemeGroupVersion.Group,
+		Version:  pd.GetObjectKind().GroupVersionKind().Version,
+		Resource: util.PlacementDecisionResource,
+	}).Update(context.Background(), unstructuredPlacementDecision, metav1.UpdateOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			_, err = c.dynamicClient.Resource(schema.GroupVersionResource{
+				Group:    v1alpha1.SchemeGroupVersion.Group,
+				Version:  pd.GetObjectKind().GroupVersionKind().Version,
+				Resource: util.PlacementDecisionResource,
+			}).Create(context.Background(), unstructuredPlacementDecision, metav1.CreateOptions{})
+			if err != nil {
+				return fmt.Errorf("failed to create placement decision: %w", err)
+			}
+
+			c.logger.Info("created placement decision", "name", pd.GetName())
+		} else {
+			return fmt.Errorf("failed to update placement decision: %w", err)
+		}
+	}
+
+	c.logger.Info("updated placement decision", "name", pd.GetName())
+	return nil
+}
+
+func (c *Controller) listPlacementDecisions() ([]runtime.Object, error) {
+	pdLister := c.listers[util.KeyForGroupVersionKind(v1alpha1.SchemeGroupVersion.Group,
+		v1alpha1.SchemeGroupVersion.Version, util.PlacementDecisionKind)]
+	if pdLister == nil {
+		return nil, fmt.Errorf("could not get lister for placememt-decision")
+	}
+
+	list, err := pdLister.List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+	return list, nil
+}
+
+func runtimeObjectToPlacementDecision(obj runtime.Object) (*v1alpha1.PlacementDecision, error) {
```

Comment:
Since this does not work for _any_ `runtime.Object` that represents a PlacementDecision, but only ones using `*unstructured.Unstructured`, I would name this func more specifically. Maybe `unstructuredRuntimeObjectToPlacementDecision`. But more likely I would do the cast from `runtime.Object` to `*unstructured.Unstructured` higher in the stack, leaving this func only to convert from `*unstructured.Unstructured` to `PlacementDecision`.